pkg/volume/storageos/storageos.go

```diff
@@ -0,0 +1,786 @@
+		return err
+	}
+
+	// Unmount from plugin's disk global mount dir.
+	err = b.TearDownAt(globalPDPath)
+	if err != nil {
+		glog.Errorf("Detach failed during unmount: %v", err)
+		return err
+	}
+
+	// Detach loop device
+	err = b.manager.DetachVolume(b, devicePath)
+	if err != nil {
+		glog.Errorf("Detach device %s failed for volume %s: %v", devicePath, b.pvName, err)
+		return err
+	}
+
+	glog.V(4).Infof("Successfully unmounted StorageOS volume %s and detached devices", b.pvName)
+
+	return nil
+}
+
+// Unmounts the bind mount, and detaches the disk only if the PD
+// resource was the last reference to that disk on the kubelet.
+func (b *storageosUnmounter) TearDownAt(dir string) error {
+	if err := util.UnmountPath(dir, b.mounter); err != nil {
+		glog.V(4).Infof("Unmounted StorageOS volume %s failed with: %v", b.pvName, err)
+	}
+	if err := b.manager.UnmountVolume(b); err != nil {
+		glog.V(4).Infof("Mount reference for volume %s could not be removed from StorageOS: %v", b.pvName, err)
+	}
+	return nil
+}
+
+type storageosDeleter struct {
+	*storageos
+	pvUID types.UID
+}
+
+var _ volume.Deleter = &storageosDeleter{}
+
+func (d *storageosDeleter) GetPath() string {
+	return getPath(d.podUID, d.volNamespace, d.volName, d.pvName, d.plugin.host)
+}
+
+func (d *storageosDeleter) Delete() error {
+	if err := d.manager.DeleteVolume(d); err != nil {
+		return err
+	}
+
+	// Clean up secret used to locate PV's API credentials.  The ref doesn't
+	// always exist, so don't report an error on fail.
+	if err := d.plugin.removeSecretRef(d.pvUID); err != nil {
+		glog.V(4).Infof("Did not delete secret ref: %v", err)
+	}
+	return nil
+}
+
+type storageosProvisioner struct {
+	*storageos
+	options volume.VolumeOptions
+}
+
+var _ volume.Provisioner = &storageosProvisioner{}
+
+func (c *storageosProvisioner) Provision() (*v1.PersistentVolume, error) {
+
+	// Apply ProvisionerParameters (case-insensitive). We leave validation of
+	// the values to the cloud provider.
+	for k, v := range c.options.Parameters {
+		switch strings.ToLower(k) {
+		case "secretname":
+			c.secretName = v
+		case "volumenamespace":
+			c.volNamespace = v
+		case "description":
+			c.description = v
+		case "pool":
+			c.pool = v
+		case "fstype":
+			c.fsType = v
+		default:
+			return nil, fmt.Errorf("invalid option %q for volume plugin %s", k, c.plugin.GetPluginName())
+		}
+	}
+
+	// Set from PVC
+	c.podNamespace = c.options.PVC.Namespace
+	c.volName = c.options.PVName
+	if c.volNamespace == "" {
+		c.volNamespace = c.options.PVC.Namespace
+	}
+	c.labels = make(map[string]string)
+	for k, v := range c.options.PVC.Labels {
+		c.labels[k] = v
+	}
+	capacity := c.options.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
+	c.size = int(volume.RoundUpSize(capacity.Value(), 1024*1024*1024))
+
+	// Store API secret location for lookup during unmount/delete.
+	secretRef, err := c.plugin.storeSecretRef(c.podNamespace, c.secretName, c.options.PVC.UID)
```

Comment:
creating secrets in every namespace isn't appropriate... if you're going to denormalize, I'd rather see the secret name and namespace specified in the PV spec (though that means reworking https://github.com/kubernetes/kubernetes/blob/master/pkg/api/persistentvolume/util.go#L26 to pass secret name and namespace to the visitor, since the storageos volume source would differ from the other sources)